ai/anthropic/v0/main.go

```diff
@@ -0,0 +1,180 @@
+
+	"github.com/instill-ai/component/base"
+)
+
+const (
+	textGenerationTask = "TASK_TEXT_GENERATION_CHAT"
+	cfgAPIKey          = "api_key"
+	host               = "https://api.anthropic.com"
+	messagesPath       = "/v1/messages"
+)
+
+var (
+	//go:embed config/definition.json
+	definitionJSON []byte
+	//go:embed config/setup.json
+	setupJSON []byte
+	//go:embed config/tasks.json
+	tasksJSON []byte
+
+	once sync.Once
+	comp *component
+)
+
+type component struct {
+	base.Component
+}
+
+func Init(bc base.Component) *component {
+	once.Do(func() {
+		comp = &component{Component: bc}
+		err := comp.LoadDefinition(definitionJSON, setupJSON, tasksJSON, nil)
+		if err != nil {
+			panic(err)
+		}
+	})
+	return comp
+}
+
+type execution struct {
+	base.ComponentExecution
+	execute func(*structpb.Struct) (*structpb.Struct, error)
+}
+
+func (c *component) CreateExecution(sysVars map[string]any, setup *structpb.Struct, task string) (*base.ExecutionWrapper, error) {
+	e := &execution{
+		ComponentExecution: base.ComponentExecution{Component: c, SystemVariables: sysVars, Task: task, Setup: setup},
+	}
+
+	// A simple if statement would be enough in a component with a single task.
+	// If the number of task grows, here is where the execution task would be
+	// selected.
+	switch task {
+	case textGenerationTask:
+		e.execute = e.generateText
+	default:
+		return nil, fmt.Errorf("unsupported task")
+	}
+	return &base.ExecutionWrapper{Execution: e}, nil
+}
+
+func (e *execution) Execute(_ context.Context, inputs []*structpb.Struct) ([]*structpb.Struct, error) {
+	outputs := make([]*structpb.Struct, len(inputs))
+
+	// An execution  might take several inputs. One result will be returned for
+	// each one of them, containing the execution output for that set of
+	// parameters.
+	for i, input := range inputs {
+		output, err := e.execute(input)
+		if err != nil {
+			return nil, err
+		}
+
+		outputs[i] = output
+	}
+
+	return outputs, nil
+}
+
+func retriveMessageContent(contentsPbValue *structpb.ListValue) []content {
+	contents := []content{}
+	for _, contentPbValue := range contentsPbValue.GetValues() {
+		contentType := contentPbValue.GetStructValue().Fields["type"].GetStringValue()
+		// anthrothpic models does not support image urls
+		if contentType == "text" {
+			content := content{
+				Type: "text",
+				Text: contentPbValue.GetStructValue().Fields["text"].GetStringValue(),
+			}
+			contents = append(contents, content)
+		}
+	}
+	return contents
+}
+
+func retriveChatMessage(chatHistoryPbList *structpb.ListValue) []message {
+	messages := []message{}
+
+	for _, messagePbValue := range chatHistoryPbList.GetValues() {
+		contents := retriveMessageContent(messagePbValue.GetStructValue().Fields["content"].GetListValue())
+		completeMessage := message{Role: messagePbValue.GetStructValue().Fields["role"].GetStringValue(), Content: contents}
+		messages = append(messages, completeMessage)
+	}
+	return messages
+}
```

Comment:
Question
Sorry, I may miss something. Please correct me if I am wrong.
Is it possible we do the flow like below?

1. Get the messagePbValue first
2. Concurrently run the slice of messagePbValue to unpack the content in the message.

I think it will be better for time complexity.
I guess it could damage the performance with the current code with 3 for loop.

cc'd @donch1989 
I am also not a master in this field. 
Please take a look the comment and feel free to give feedback to us.

Reply:
I've wrote a multi-threaded version. However, the main bottleneck should be the respond time of AI vendor, thus this modification won't make a lot of difference.

Reply:
> However, the main bottleneck should be the respond time of AI vendor

Agree on this point.
But, they are separated to be improved.
1. AI vendor can improve their response time on their responsibility.
2. VDP can improve VDP's process time on VDP's responsibility.

It would become a bottleneck if the response message becomes huge. 
e.g. 1000000 messages contain 100000 content.

However, I am not sure if it will happen in the future that short time can respond a huge piece of messages.
If it could happen, we should do multi-threaded version for VDP responsibility in my opinion.
What do you think about this topic?

Reply:
My current implementation processes each message concurrently and parses the contents of each message sequentially. This approach assumes that the number of messages is usually greater than the number of contents each message stores, which is typically the case in mainstream AI applications.

Here's the polished sentence:

As for the second part, in my experience, the functionality that needs to be implemented using multi-threading is either (a) I/O intensive, such as web scraping, or (b) computationally intensive, such as machine learning algorithms. For simple 2D/3D array operations, based on my experience in competitive programming [reference](https://www.geeksforgeeks.org/knowing-the-complexity-in-competitive-programming/), a typical modern processor can perform 10^8 simple operations per second. I use this as a guideline to determine whether to implement multi-threading. For example, iterating through an array of 10^4 elements will take roughly 10^-4 seconds. Splitting the workload across 8 physical cores would take around 10^-5 seconds, so the time saved might not justify the engineering effort required to implement multi-threading.

Reply:
@namwoam 

> competitive programming [reference](https://www.geeksforgeeks.org/knowing-the-complexity-in-competitive-programming/)

Thanks for sharing this good reference with me.

> So the key point is to know the common usecase/workload of our components to better optimize the efficiency of VDP.

I do not know the AI model deeply.
Could you help us research it and judge the possible use case/workload?
Please also share some references for us to let us know this information as well.
Thank you! 🙇

Reply:
Here's a polished version of the message:

According to an IBM blog post [link](https://www.ibm.com/blog/generative-ai-use-cases/), common use cases for generative AI can be categorized into eight areas:

1. Code generation
2. Product development
3. Sales and marketing
4. Project management and operations
5. Graphic design and video
6. Business and employee management
7. Customer support and customer service
8. Fraud detection and risk management

Among these, "Project management and operations" and "Fraud detection and risk management" are likely to benefit most from long-term memory capabilities, as they require historical context to generate useful insights.

However, for such use cases, enterprises might prefer a Retrieval-Augmented Generation (RAG) system to quickly retrieve relevant context and inject it into the model's prompt. This approach is more economically viable, given that the computational complexity of Large Language Models (LLMs) is roughly O(N^2) relative to prompt length (due to the attention mechanism in the Transformer architecture). RAG allows for more efficient use of resources compared to directly feeding all data into the model.

Reply:
Briefly, you are saying that no use case in the real world will not have the big workload that concerns us. right?

>  so the time saved might not justify the engineering effort required to implement multi-threading.

And, I'd like to confirm the reason why we do not improve the time complexity here is engineering effort.
Is there other specific reason?

Please feel free to correct me if I misunderstand.

Reply:
Regarding component throughput bottlenecks, I believe we'll only need to be concerned when AI vendors support network-intensive tasks like video-to-text conversion or video generation.
As for multi-threading, it's inherently complex due to race conditions [simple tutorial](https://www.geeksforgeeks.org/race-condition-vulnerability/). Controlling the locking and unlocking of different processes is a challenging task that requires significant engineering effort.

Reply:
Last time, I met the similar context in Slack component.
I did put a bit effort on dealing with multi-threading code. 
But, I feel it is not my bottleneck in terms of engineering effort.

However, I may miss considering some points that you have known because you have done so many research.
If it does makes sense for you now, I think you can decide it for this conversation.
Simply put, it LGTM!